Add Soviet10

Smittytron opened the pull request.

Smittytron committed:
Add Soviet10

pchote commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,269 @@
+	            if #guards == 0 then
+					RunForIt()
+					return
+				end
```
Indentation problems here.

pchote commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,269 @@
+	            end
+
+	            if #guards == 0 then
+					RunForIt()
```
IMO it would be good to have a comment here to make it clear that this runs on the first truck, then the return below means that we don't loop over the rest.

pchote commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,269 @@
+
+	Trigger.AfterDelay(DateTime.Seconds(1), function()
+		Reinforcements.Reinforce(USSR, ConvoyEscort, { ConvoyEntry.Location, TanksStop.Location })
+		Media.PlaySpeechNotification(USSR, "ReinforcementsArrived")
```
The original mission script used the specific "Convoy Approaching" audio.

pchote reviewed (commented):
The original mission script (see https://gist.githubusercontent.com/pchote/558172f2bb290915e0d506cca6b99b8f/raw/f1dd9c18d9ab37b7cd82453c2f772a726fe5eb47/scu10ea.txt, mostly complete except for the missions assigned to spawned teams) shows a few extra triggers to destroy buildings when barrels are blown up, granting map vision when capturing the radar dome (not sure this makes sense to port over, but maybe we could do something else?), and selling off all structures when certain actors are attacked.

Do we want to include these?

pchote commented:
It would be nice polish IMO if we could script the convoy trucks to drive off the map before triggering mission accomplished.

pchote commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,269 @@
+end
+
+RunForIt = function()
+	Media.PlaySoundNotification(USSR, "AlertBleep")
```
ATM we play the notification and show the message repeatedly once this triggers.
We should instead have a flag to stop `TrucksFollow` from being requeued, and then probably also use an OnIdle trigger to make sure the trucks keep trying to move to the exit if they get nudged or otherwise interrupted.

Smittytron commented:
> It would be nice polish IMO if we could script the convoy trucks to drive off the map before triggering mission accomplished.

Done

> ...destroy buildings when barrels are blown up, granting map vision when capturing the radar dome (not sure this makes sense to port over, but maybe we could do something else?), and selling off all structures when certain actors are attacked.

Added a kill trigger on the first AA gun. It looks like most of those kill triggers are on trees to turn them into dead trees when the barrels around them are destroyed. As for the radar, a big difference in this mission is OpenRA's planes can clear shroud, so I don't think the player is hurting for map info.

Selling is an interesting point. We haven't really used `Sell()` in any missions so far but I know several buildings in the OG would sell on a damage threshold. Judging by the cells I'm guessing the sell trigger is to sell the radar dome, but I can't tell for sure without re-importing because I deemed the GoodGuy player unnecessary and changed those actors to Greece. We can use `Sell()` if you want here but I'd want to start adding it in every mission old and new as it will be random to just have it here.

I also added a secondary objective to keep all trucks alive.

darcyrush commented:
Plays ok on hard but some strange follow behavior of one of the convoy trucks near the end almost cost me the mission. And the trucks didn't drive off the map for me

[ra-2021-02-27T085004Z.zip](https://github.com/OpenRA/OpenRA/files/6054026/ra-2021-02-27T085004Z.zip)

pchote commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,297 @@
+		if a.Type == "truk" then
+			a.Stop()
+			Trigger.OnIdle(a, function()
+				a.Move(Exit.Location)
```
Is this going to fight with `TrucksFollow` for control?

Smittytron replied:
Judging by the comment from Darcyrush it must be.

pchote commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,297 @@
+		Media.DisplayMessage(p.GetObjectiveDescription(id), "New " .. string.lower(p.GetObjectiveType(id)) .. " objective")
+	end)
+
+	EscortTrucks = USSR.AddObjective("Get the convoy across the map.")
```
How about "Escort the convoy though the mountain pass"?

pchote commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,297 @@
+	end)
+
+	EscortTrucks = USSR.AddObjective("Get the convoy across the map.")
+	ProtectEveryTruck = USSR.AddObjective( "Don't lose a single truck.", "Secondary", false)
```
Stray space between `( "`.

Would it be too brutal to instead have a "Convoy arrives at the destination unharmed" objective where they aren't allowed to be damaged at all?

Smittytron replied:
> Would it be too brutal...

I think it would be considering the player doesn't have complete control of where the trucks end up and any sort of splash damage (i.e. barrels) would be a frustrating way to lose an objective.

pchote commented:
⚠️ Rant incoming ⚠️ TLDR: i'm not actually fussed about what we do here right now, but would really like if we could start planning for a future where scripts and events are consistent throughout the campaign.

> It looks like most of those kill triggers are on trees to turn them into dead trees when the barrels around them are destroyed.

> Selling is an interesting point. 

These both touch on my main motiviation for developing the mission script decoder: I expect that some of our mission ports are missing events/strategies because the person porting them didn't know that they existed, or didn't get the details of the implementation quite right. We also have inconsistencies in how we handle base reconstruction / fire sales, harvester defense behaviour (https://i.redd.it/e35reiuyhlc51.jpg), and so on which IMO makes our campaign implementation feel less coherant than it should.

In the short to medium term I would really like if we could use the new tools to understand exactly how the originals worked, and write a comment at the top of each mission's lua file that explains the original mission script and our intentional and unintentional deviations from it. This information is then available for when we hopefully one day go through and properly implement our consistent "OpenRA style" mission adjustments for aircraft vision, mcv undeploy, engineers, hinds, and so on.

I'll repeat again that I don't think we should be aiming for 1:1 script compatibility, but rather that all differences are deliberately decided and documented. I think its a bit sad if we end up missing details no real reason except that we didn't know about them or couldn't be bothered to add them.

The radar dome is a good example here, I dug into this properly after your comment and confirmed that the AI will sell it (the dome is the only structure owned by goodguy) if the player crosses the red line or kills any of the circled units:

<img width="568" alt="Screenshot 2021-02-27 at 13 48 02" src="https://user-images.githubusercontent.com/167819/109390806-95d07580-790b-11eb-951c-dce077a8b0d7.png">

This is a nice little detail because it means that you need to be careful to not push forward too early if you want to get the benfits from the dome.

Another example in this mission are the two rows of turrets at the pinch point. The original map scripted the AI to rebuild these if they are destroyed, and used the trp3-6 and det3-6 triggers to script destroying them if you shoot the barrels.

I'm not sure we want to keep the detonation triggers here, since the aircraft vision otherwise makes this mission really easy. But we should document this in the script, and should remove the barrels from the base since they are obviously there purely for this purpose so it will be confusing to players if they find that this strategy no longer works. IMO an even better solution would be to adjust these mini bases to have two sets of two AA guns with barrels between them, so the player must use their ground units to clear them before they can move the Migs in to take out the turrets.

I found that the remastered map editor does a good job at displaying things with special behaviours attached (green = triggers, red = auto rebuild), but needed to get https://www.ppmforums.com/topic-54754/improvements-for-the-map-editor/ before it would load the soviet10 ini. This plus my utility command together will hopefully make it reasonably easy to understand the original mission scripts.

<img src="https://user-images.githubusercontent.com/167819/109391875-3a08eb00-7911-11eb-8eec-0d9e4b45fe7d.png">

abcdefg30 commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,318 @@
+	Utils.Do(Airfields, function(runway)
+		Trigger.OnProduction(runway, function(a, b)
+			if b.Type == "mig" then
+				Migs[#Migs + 1] = b
```
```suggestion
				Trigger.OnKilled(b, function()
					USSR.MarkFailedObjective(SaveMigs)
				end)
```

Smittytron commented:
Updated with a secondary objective for keeping migs alive. Added the chinook drop that would ambush the player below the service depot. And to address the point about documentation I added notes at the top for the things I've left out that I know are in the original.

KautionORA commented:
Just finished this on hard, didn't come across anything in particular, except for a weird pathing bug causing the truck to drive into the enemy and commits suicide. 

the only other thing worth noting is that the secondary objectives dont show they're completed once the main objective is completed, which i assume should be the trigger for completing the secondary objectives besides from the actual goal.

[Kaution files Soviet 10.zip](https://github.com/OpenRA/OpenRA/files/6272676/Kaution.files.Soviet.10.zip)

pchote commented:
The mission script has a case that is supposed to mark "Do not lose a single truck" as completed, so it may be a bug if that isn't triggering. The script doesn't have any code to mark the mig objective complete, so that does appear to be an oversight.

Smittytron commented:
Updated. Added the logic for completing the mig objective. The replay provided showed a failed objective for saving all trucks. I got all objectives to complete for me in another playthrough.

pchote commented on mods/ra/maps/soviet-10/rules.yaml:
```diff
@@ -0,0 +1,100 @@
+	LuaScript:
+		Scripts: campaign-global.lua, soviet10.lua
+	MissionData:
+		Briefing: You must defend a Soviet convoy is moving through Allied occupied territory. Using the new MIG jet and a compliment of Yaks, get the convoy through the area intact.\n\nBe careful -- your resources for this mission are very limited. If at least one truck makes it through to the other side, the mission will be a success.
```
Typo: missing "that" in "convoy that is moving through"

pchote commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,349 @@
+MissionSetup = function()
+	Trigger.OnEnteredFootprint(WinTriggerArea, function(a, id)
+		if a.Type == "truk" then
+			--stop truck from looping through TrucksFollow
```
Style nit: Comments should have a space and start with a capital letter (`-- Stop truck...`)

pchote reviewed (dismissed):
LGTM. Just two very minor nits, and its awkward that the planes take off and then re-land at the start but fixing that is out of scope of this PR.

Smittytron commented:
Updated

reaperrr commented on mods/ra/maps/soviet-10/soviet10.lua:
```diff
@@ -0,0 +1,349 @@
+		Media.DisplayMessage(p.GetObjectiveDescription(id), "New " .. string.lower(p.GetObjectiveType(id)) .. " objective")
+	end)
+
+	EscortTrucks = USSR.AddObjective("Escort the convoy though the mountain pass.")
```
th _r_ ough

reaperrr commented:
LGTM after that.

Smittytron commented:
Updated

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ce3669b0e1413527e261f5c820aa3b968eb659df...2e106fef0877e35c9b2a31aca6bd05757eb05420
